src/values/set.ts

```diff
@@ -0,0 +1,181 @@
+import { Client } from "../client";
+import { Query, fql } from "../query-builder";
+import { QuerySuccess, QueryValue } from "../wire-protocol";
+
+/**
+ * A materialize view of a Set.
+ * @see {@link https://fqlx-beta--fauna-docs.netlify.app/fqlx/beta/reference/language/types#set}
+ */
+export class Page<T extends QueryValue> {
+  /** A materialized page of data */
+  readonly data: T[];
+  /**
+   * A pagination cursor, used to obtain additional information in the Set.
+   * If `after` is not provided, then `data` must be present and represents the
+   * last Page in the Set.
+   */
+  readonly after?: string;
+
+  constructor({ data, after }: { data: T[]; after?: string }) {
+    this.data = data;
+    this.after = after;
+  }
+}
+
+/**
+ * A un-materialize Set. Typically received when a materialized Set contains
+ * another set, the EmbeddedSet does not contain any data to avoid potential
+ * issues such as self-reference and infinite recursion
+ * @see {@link https://fqlx-beta--fauna-docs.netlify.app/fqlx/beta/reference/language/types#set}
+ */
+export class EmbeddedSet {
+  /**
+   * A pagination cursor, used to obtain additional information in the Set.
+   */
+  readonly after: string;
+
+  constructor(after: string) {
+    this.after = after;
+  }
+}
+
+/**
+ * A class to provide an iterable API for fetching multiple pages of data, given
+ * a Fauna Set
+ */
+export class SetIterator<T extends QueryValue>
```

Comment:
We should give users an optional page size control.

Reply:
Good idea! Current caveat is that the first page will always be up to 16. We can fix eventually by adding `.paginate(${size})` to the end of their query, in the case they use `client.paginate` or otherwise use `SetIterator.fromQuery`, but the `<set>.paginate` method doesn't yet return `@set`-tagged values.

Reply:
hmmmm. should we superficially just return groups of the specified size, regardless if the result is larger (should only be necessary for the first page)

Reply:
Fair enough. Adding in size is additive, so it's certainly safe to hold off if anything. I think it is also fair to require users to use `.paginate(number)` directly (once core makes that return the tagged value).

Reply:
yeah i think if core handled it it would be a simpler API as we are just using core then and not doing any shimming.